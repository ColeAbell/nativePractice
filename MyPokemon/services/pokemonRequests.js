const apiUrl = 'https://pokeapi.co/api/v2/pokemon/';

export async function getAllPokemon(url = apiUrl) {
  const results = await fetch(url);
  if (results.status === 200) {
    return await results.json();
  }
  throw new Error('Did not catch them all');
}

export async function getOnePokemon(url) {
  const results = await fetch(url);
  if (results.status === 200) {
    return await results.json();
  }
  throw new Error('Did not catch them all');
}
