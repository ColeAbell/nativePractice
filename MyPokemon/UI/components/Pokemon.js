import {Image, Pressable, StyleSheet, Text, View, Platform} from 'react-native';
import React, {useEffect, useState} from 'react';
import {getOnePokemon} from '../../services/pokemonRequests';

export default function Pokemon({details, press}) {
  const [sprite, setSprite] = useState({sprite: null, allDetails: {}});

  async function getSprite() {
    try {
      const info = await getOnePokemon(details.url);
      setSprite({sprite: info.sprites['front_default'], allDetails: info});
    } catch (e) {
      console.error(e);
    }
  }

  function handlePress() {
    press(sprite.allDetails);
  }

  useEffect(() => {
    getSprite();
  }, []);

  return (
    <View style={styles.container}>
      <Pressable onPress={handlePress}>
        <View style={styles.innerContainer}>
          <Text style={styles.name}>{details.name}</Text>
          <Image source={{uri: sprite.sprite}} style={styles.sprite} />
        </View>
      </Pressable>
    </View>
  );
}

const styles = StyleSheet.create({
  sprite: {width: 150, height: 150},
  container: {
    width: 180,
    height: 200,
    margin: 8,
    backgroundColor: '#7fffd4',
    borderRadius: 20,
    borderWidth: 6,
  },
  innerContainer: {
    alignItems: 'center',
  },
  name: {
    marginTop: 10,
    fontStyle: 'italic',
    fontSize: 25,
  },
});
