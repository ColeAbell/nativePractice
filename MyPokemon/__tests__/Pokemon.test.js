import ditto from './ditto.json';
import {
  render,
  screen,
  fireEvent,
  act,
  waitFor,
} from '@testing-library/react-native';
import 'react-native';
import React from 'react';
import Pokemon from '../UI/components/Pokemon';
import * as pokereq from '../services/pokemonRequests';

global.fetch = require('jest-fetch-mock');

beforeEach(() => {
  fetch.resetMocks();
});

/*
test('returns result if object', () => {
  fetch.mockResponseOnce(JSON.stringify(ditto));
  const onResponse = jest.fn();
  const onError = jest.fn();
  return getOnePokemon()
    .then(onResponse)
    .catch(onError)
    .finally(() => {
      expect(onResponse).toHaveBeenCalled();
      expect(onResponse.mock.calls[0][0]).toEqual(ditto);
      console.log(onResponse.mock.calls);
    });
});
*/

test('Does Pokemon render correctly', async () => {
  //fetch.mockResponseOnce(JSON.stringify(ditto));
  console.log(pokereq);
  /*
  jest.spyOn(pokereq, 'getOnePokemon').mockImplementation(() => {
    return JSON.stringify(ditto);
  });
  */
  pokereq.getOnePokemon = jest.fn().mockImplementation(url => {
    console.log(url);
    return ditto;
  });
  //jest.useFakeTimers();
  const handlePress = jest.fn();
  //async function wait() {
  const tree = render(
    <Pokemon
      details={{
        name: 'ditto',
        url: 'https://pokeapi.co/api/v2/pokemon/1/',
      }}
      press={handlePress}
    />,
  );

  //}
  //await wait();
  /*
    await act(async () => {
      jest.runAllTimers();
    });
    */
  expect(screen.getByTestId('activity')).toBeTruthy();
  expect(pokereq.getOnePokemon).toHaveBeenCalled();

  await waitFor(() => {
    expect(screen.getByText('ditto')).toBeTruthy();
  });
  expect(screen.getByTestId('sprite').props.source.uri).toEqual(
    'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/132.png',
  );

  fireEvent.press(await screen.findByTestId('pressable'));
  expect(handlePress).toBeCalledWith(ditto);
  expect(pokereq.getOnePokemon).toBeCalled();
  expect(tree).toMatchSnapshot();
});
