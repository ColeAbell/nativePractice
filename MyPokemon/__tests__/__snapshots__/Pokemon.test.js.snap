// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Does Pokemon render correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#7fffd4",
      "borderRadius": 20,
      "borderWidth": 6,
      "height": 200,
      "margin": 8,
      "width": 180,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="pressable"
  >
    <View
      style={
        {
          "alignItems": "center",
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 25,
            "fontStyle": "italic",
            "marginTop": 10,
          }
        }
      >
        ditto
      </Text>
      <Image
        source={
          {
            "uri": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/132.png",
          }
        }
        style={
          {
            "height": 150,
            "width": 150,
          }
        }
        testID="sprite"
      />
    </View>
  </View>
</View>
`;

exports[`Does home screen work 1`] = `
<View
  style={
    {
      "backgroundColor": "#7fffd4",
      "borderRadius": 20,
      "borderWidth": 6,
      "height": 200,
      "margin": 8,
      "width": 180,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="pressable"
  >
    <View
      style={
        {
          "alignItems": "center",
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 25,
            "fontStyle": "italic",
            "marginTop": 10,
          }
        }
      >
        ditto
      </Text>
      <Image
        source={
          {
            "uri": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/132.png",
          }
        }
        style={
          {
            "height": 150,
            "width": 150,
          }
        }
        testID="sprite"
      />
    </View>
  </View>
</View>
`;
